packages/flutter/lib/src/widgets/modal_button_sheet.dart

```diff
@@ -0,0 +1,652 @@
+      // When using WillPop, animation doesn't end at 1.
+      // Check more in detail the problem
+      if (!widget.animationController.isCompleted) {
+        widget.animationController.value = 1;
+      }
+    });
+    _bounceDragController.reverse();
+  }
+
+  bool _isCheckingShouldClose = false;
+
+  FutureOr<bool> shouldClose() async {
+    if (_isCheckingShouldClose) return false;
+    if (widget.shouldClose == null) return false;
+    _isCheckingShouldClose = true;
+    final result = await widget.shouldClose!();
+    _isCheckingShouldClose = false;
+    return result;
+  }
+
+  late ParametricCurve<double> animationCurve;
+
+  void _handleDragUpdate(double primaryDelta) async {
+    animationCurve = Curves.linear;
+    assert(widget.enableDrag, 'Dragging is disabled');
+
+    if (_dismissUnderway) return;
+    isDragging = true;
+
+    final progress = primaryDelta / (_childHeight ?? primaryDelta);
+
+    if (widget.shouldClose != null && hasReachedWillPopThreshold) {
+      _cancelClose();
+      final canClose = await shouldClose();
+      if (canClose) {
+        _close();
+        return;
+      } else {
+        _cancelClose();
+      }
+    }
+
+    // Bounce top
+    final bounce = widget.bounce == true;
+    final shouldBounce = _bounceDragController.value > 0;
+    final isBouncing = (widget.animationController.value - progress) > 1;
+    if (bounce && (shouldBounce || isBouncing)) {
+      _bounceDragController.value -= progress * 10;
+      return;
+    }
+
+    widget.animationController.value -= progress;
+  }
+
+  void _handleDragEnd(double? velocity) async {
+    assert(widget.enableDrag, 'Dragging is disabled');
+
+    animationCurve = BottomSheetSuspendedCurve(
+      widget.animationController.value,
+      curve: _defaultCurve,
+    );
+
+    if (_dismissUnderway || !isDragging) return;
+    isDragging = false;
+    _bounceDragController.reverse();
+
+    var canClose = true;
+    if (widget.shouldClose != null && hasReachedWillPopThreshold) {
+      _cancelClose();
+      canClose = await shouldClose();
+    }
+    if (canClose) {
+      // If speed is bigger than _minFlingVelocity try to close it
+      if (velocity != null && velocity > _minFlingVelocity) {
+        _close();
+      } else if (hasReachedCloseThreshold) {
+        if (widget.animationController.value > 0.0) {
+          widget.animationController.fling(velocity: -1.0);
+        }
+        _close();
+      } else {
+        _cancelClose();
+      }
+    } else {
+      _cancelClose();
+    }
+  }
+
+  // As we cannot access the dragGesture detector of the scroll view
+  // we can not know the DragDownDetails and therefore the end velocity.
+  // VelocityTracker it is used to calculate the end velocity  of the scroll
+  // when user is trying to close the modal by dragging
+  VelocityTracker? _velocityTracker;
+  DateTime? _startTime;
+
+  void _handleScrollUpdate(ScrollNotification notification) {
```

Comment:
having all these touch handling stuff is understandable but I think they can just be in the cupertino package. For prior art, see all the back gesture handling and how they're all re-used by the MaterialPageRoute as well through the PageTransitionsTheme.